criu/proc_parse.c

```diff
@@ -206,6 +208,20 @@ static void parse_vma_vmflags(char *buf, struct vma_area *vma_area)
 
 	if (vma_area->e->madv)
 		vma_area->e->has_madv = true;
+
+	/*
+	 * We set MAP_PRIVATE flag on vma_area->e->flags right after parsing
+	 * a first line of VMA entry in /proc/<pid>/smaps file:
+	 * 7fa84fa70000-7fa84fa95000 rw-p 00000000 00:00 0
+	 * but it's too early and we can't distinguish between MAP_DROPPABLE
+	 * and MAP_PRIVATE mappings yet, as they both private mappings in nature
+	 * and at this point we haven't yet read "VmFlags:" line in smaps.
+	 *
+	 * Let's detect this situation and drop MAP_PRIVATE flag while keep
+	 * MAP_DROPPABLE, otherwise restorer's restore_mapping() helper will fail.
+	 */
+	if ((vma_area->e->flags & MAP_PRIVATE) && (vma_area->e->flags & MAP_DROPPABLE))
```

Comment:
do we need to dump content of MAP_DROPPABLE areas? The kernel commit message says:
```
b) Under memory pressure, mm can just drop the pages (so that they're
   zero when read back again).
```

Reply:
yes I think we can drop the entire contents of them, but what is more important is that right now (without these fixes) we restore `MAP_DROPPABLE` mappings just as `MAP_PRIVATE` ones and lose `VM_WIPEONFORK`. This can be fatal if application relies on the fact that after `fork()` mapping should be null-ified, because it won't be. I would classify this as a memory corruption-like issue.

Reply:
Have fixed this. Now we drop `MAP_DROPPABLE` mappings content after C/R.

Thanks!